presto-orc/src/main/java/com/facebook/presto/orc/StripeReader.java

```diff
@@ -514,26 +511,20 @@ static boolean isIndexStream(Stream stream)
         return bloomFilters.build();
     }
 
-    private Map<StreamId, List<RowGroupIndex>> readColumnIndexes(Map<StreamId, Stream> streams, Map<StreamId, OrcInputStream> streamsData, Map<Integer, List<HiveBloomFilter>> bloomFilterIndexes)
+    private Map<StreamId, List<RowGroupIndex>> readColumnIndexes(Map<StreamId, Stream> streams, Map<StreamId, OrcInputStream> streamsData, StripeId stripeId)
             throws IOException
     {
+        // read the bloom filter for each column
+        Map<Integer, List<HiveBloomFilter>> bloomFilterIndexes = readBloomFilterIndexes(streams, streamsData);
+
         ImmutableMap.Builder<StreamId, List<RowGroupIndex>> columnIndexes = ImmutableMap.builder();
         for (Entry<StreamId, Stream> entry : streams.entrySet()) {
+            StreamId streamId = entry.getKey();
             Stream stream = entry.getValue();
             if (stream.getStreamKind() == ROW_INDEX) {
-                OrcInputStream inputStream = streamsData.get(entry.getKey());
-                List<HiveBloomFilter> bloomFilters = bloomFilterIndexes.get(entry.getKey().getColumn());
-                List<RowGroupIndex> rowGroupIndexes = metadataReader.readRowIndexes(hiveWriterVersion, inputStream);
-                if (bloomFilters != null && !bloomFilters.isEmpty()) {
-                    ImmutableList.Builder<RowGroupIndex> newRowGroupIndexes = ImmutableList.builder();
-                    for (int i = 0; i < rowGroupIndexes.size(); i++) {
-                        RowGroupIndex rowGroupIndex = rowGroupIndexes.get(i);
-                        ColumnStatistics columnStatistics = rowGroupIndex.getColumnStatistics()
-                                .withBloomFilter(bloomFilters.get(i));
-                        newRowGroupIndexes.add(new RowGroupIndex(rowGroupIndex.getPositions(), columnStatistics));
-                    }
-                    rowGroupIndexes = newRowGroupIndexes.build();
-                }
+                OrcInputStream inputStream = streamsData.get(streamId);
+                List<HiveBloomFilter> bloomFilters = bloomFilterIndexes.get(streamId.getColumn());
+                List<RowGroupIndex> rowGroupIndexes = stripeMetadataSource.getRowIndexes(metadataReader, hiveWriterVersion, stripeId, streamId, inputStream, bloomFilters);
```

Comment:
Maybe better to add some per query counters to get some insights on the size of the row index and cache hit rate when the cache is enabled.

StripeReader has a runtimeStats which can be passed into the getRowIndexes function and we could only populate the metrics in CachingStripeMetadataSource when the cache is enabled.